packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,209 @@
+  ///
+  /// Defaults to a simple string-matching search of [options].
+  final AutocompleteSearchFunction<T> search;
+
+  /// The current results being returned by [search].
+  ///
+  /// This is a [ValueNotifier] so that UI may be updated when results change.
+  final ValueNotifier<List<T>> results = ValueNotifier<List<T>>(<T>[]);
+
+  /// Clean up memory created by the AutocompleteController.
+  ///
+  /// Call this when the AutocompleteController is no longer needed, such as in
+  // the dispose method of the widget it was created in.
+  void dispose() {
+    textEditingController.removeListener(_onQueryChanged);
+    // TODO(justinmc): Shouldn't be disposed if it wasn't created here.
+    textEditingController.dispose();
+  }
+
+  // Called when textEditingController reports a change in its value.
+  void _onQueryChanged() {
+    final List<T> resultsValue = search == null
+        ? _searchByString(textEditingController.value.text)
+        : search(textEditingController.value.text);
+    assert(resultsValue != null);
+    results.value = resultsValue;
+  }
+
+  // The default search function, if one wasn't supplied.
+  List<T> _searchByString(String query) {
+    return options
+        .where((T option) => option.toString().contains(query))
+        .toList();
+  }
+}
+
+/// A widget for helping the user to search a list of options and select a
+/// result.
+///
+/// This is a core framework widget with very basic UI. Try using [Autocomplete]
+/// or [AutocompleteCupertino] before resorting to this widget.
+class AutocompleteCore<T> extends StatefulWidget {
+  /// Create an instance of AutocompleteCore.
+  ///
+  /// [autocompleteController], [buildField], and [buildResults] must not be
+  /// null.
+  const AutocompleteCore({
+    @required this.autocompleteController,
+    @required this.buildField,
+    @required this.buildResults,
+  }) : assert(autocompleteController != null),
+       assert(buildField != null),
+       assert(buildResults != null);
+
+  /// The controller that provides access to the main autocomplete state and
+  /// logic.
+  final AutocompleteController<T> autocompleteController;
+
+  /// Builds the field that is used to input the query.
+  final AutocompleteFieldBuilder buildField;
+
+  /// Builds the selectable results of searching.
+  final AutocompleteResultsBuilder<T> buildResults;
+
+  @override
+  _AutocompleteCoreState<T> createState() =>
+      _AutocompleteCoreState<T>();
+}
+
+class _AutocompleteCoreState<T> extends State<AutocompleteCore<T>> {
+  T _selection;
+
+  void _onChangeResults() {
+    setState(() {});
+  }
+
+  void _onChangeQuery() {
+    if (widget.autocompleteController.textEditingController.value.text == _selection) {
+      return;
+    }
+    setState(() {
+      _selection = null;
+    });
+  }
+
+  void onSelected (T result) {
+    setState(() {
+      _selection = result;
+      widget.autocompleteController.textEditingController.text = result.toString();
```

Comment:
is `T.toString()`-ing non-string types a normal thing for the `TextEditingController`? I notice you did that here and with the `_searchByString`, and I'm wondering how this actually works out in practice.

Reply:
I'm thinking that T will often be some app-specific type like a User object, and if the developer wants Autocomplete to search their Users automatically, then they need to provide a sensible toString method.  If their search logic is even more complicated than that, then they can pass a custom search function.

Actually I just added an example of this to the AutocompleteCore docs because you're right that it's not immediately clear how to do this.  Let me know it makes sense to you after looking at the example.

Reply:
I also ended up adding the onSelected parameter to allow users to customize that. It's useful when using a custom type in case you want to populate the field with something different than toString.  It's also in the example.